src/contrib/persistence/Akka.Persistence.Sql.Common/Journal/BatchingSqlJournal.cs

```diff
@@ -1388,29 +1389,111 @@ protected void AddParameter(TCommand command, string paramName, DbType dbType, o
         /// <param name="param">Parameter to customize</param>
         protected virtual void PreAddParameterToCommand(TCommand command, DbParameter param) { }
         
+        /// <summary>
+        /// Select the buffer that has the smallest id on its first item, retrieve a maximum Setup.MaxBatchSize
```

Comment:
Because this indicates the "oldest" outstanding operation?

Reply:
Yes, it is "roughly first in, first out". Question is, how sensitive is journal request ordering between read and write operations, are they strict or can operations happen in asynchrony?
If they're tightly coupled, I'd have to change this logic so that only consecutive requests of the same operations are allowed to be batched, ie. all queries, all writes, and all deletes.

Reply:
This is actually a big question that needs to be addressed.

Reply:
FWIW, I know that persistence-jdbc makes a point of blocking `ReadHighestSequenceNrAsync` for a given PersistenceId ([example in ported code](https://github.com/akkadotnet/Akka.Persistence.Linq2Db/blob/dev/src/Akka.Persistence.Sql.Linq2Db/Journal/Linq2DbWriteJournal.cs#L141-L159).

I think this is done, because the first step of Recovery in `AsyncWriteJournal` is to read the max sequence number from which to replay. This way, if somehow the PersistentActor crashes before a `PersistAsync` write completes, Recovery sequence will still read up to the correct sequence number.

Reply:
So, to make things as safe as possible, i guess i should only batch consecutive similar commands, unless it is a ReadHighestSequenceNr, in which i should stop. This is turning into an `UnBatchedSqlJournal` real fast...

Reply:
Well, all query types should be safe to be batched together as long s they're consecutive, right. The danger comes when it is interleaved with writes, and for write operations, consecutive writes and consecutive deletes are fine.

Reply:
After pondering it, I think we're probably Safe since BatchingSqlJournal is not from AsyncJournalBase and therefore doing recovery differently.

At -worst-, I think we would see a scenario where after recovery, the actor crashes b/c it didn't pick up the right sequence number. In that at worst, I can't imagine it would be much more than restarting the actor (or at worst the whole system.) 

That's still a better worst case than with current state of BatchingSqlJournal (where writes are being confirmed when they may still fail.) Under heavy loads I would see torn write behavior (i.e. missing sequence numbers in DB) basically corrupting the journal state.

Reply:
@to11mtm do you approve this set of changes then?